edb/edgeql/compiler/inference/multiplicity.py

```diff
@@ -0,0 +1,701 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+"""EdgeQL multiplicity inference.
+
+A top-down multiplicity inferer that traverses the full AST populating
+multiplicity fields and performing multiplicity checks.
+"""
+
+
+from __future__ import annotations
+from typing import *
+
+import functools
+
+from edb import errors
+
+from edb.edgeql import qltypes
+
+from edb.schema import pointers as s_pointers
+from edb.schema import constraints as s_constr
+
+from edb.ir import ast as irast
+
+from . import cardinality
+from . import context as inference_context
+
+
+ZERO = qltypes.Multiplicity.ZERO
+ONE = qltypes.Multiplicity.ONE
+MANY = qltypes.Multiplicity.MANY
+
+
+def _max_multiplicity(
+    args: Iterable[qltypes.Multiplicity]
+) -> qltypes.Multiplicity:
+    # Coincidentally, the lexical order of multiplicity is opposite of
+    # order of multiplicity values.
+    arg_list = list(args)
+    if not arg_list:
+        return ZERO
+    else:
+        return min(arg_list)
+
+
+def _common_multiplicity(
+    args: Iterable[irast.Base],
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return _max_multiplicity(
+        infer_multiplicity(a, scope_tree=scope_tree, ctx=ctx) for a in args)
+
+
+@functools.singledispatch
+def _infer_multiplicity(
+    ir: irast.Expr,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # return MANY
+    raise ValueError(f'infer_multiplicity: cannot handle {ir!r}')
+
+
+@_infer_multiplicity.register
+def __infer_none(
+    ir: None,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # Here for debugging purposes.
+    raise ValueError('invalid infer_multiplicity(None, schema) call')
+
+
+@_infer_multiplicity.register
+def __infer_statement(
+    ir: irast.Statement,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return infer_multiplicity(
+        ir.expr, scope_tree=scope_tree, ctx=ctx)
+
+
+@_infer_multiplicity.register
+def __infer_empty_set(
+    ir: irast.EmptySet,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return ZERO
+
+
+@_infer_multiplicity.register
+def __infer_type_introspection(
+    ir: irast.TypeIntrospection,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # The result is always ONE, but we still want to actually
```

Comment:
This comment seems out of place.

Reply:
This is actually one of the places where you could sneak multiplicity-breaking expression without triggering an error:
```
WITH X := Foo {bar := {Bar, Bar}}
SELECT INTROSPECT TYPEOF X.bar;
```
By the time we get to validating multiplicity, the original expression in the `TYPEOF` is lost. Technically, this is in the same category of nitpicking of whether we validate "unused" expressions, but if we want full consistency then this place should be checked as well.

I've added a longer comment to clarify.